content/post/ddos-detection.md

```diff
@@ -0,0 +1,419 @@
+
+```python
+class ServerPartitionFunction(object):
+    def __init__(self, partitions):
+        self.part_size = 2**32/partitions
+
+    def partition(self, data):
+        parts = data['server'].split('.')
+        ipnum = ((int(parts[0]) << 24) + (int(parts[1]) << 16) +
+                 (int(parts[2]) << 8) + int(parts[3]))
+        return ipnum/self.part_size
+```
+
+The state object is where things get interesting. For each server, we need to maintain a 60 seconds long rolling window of aggregate data. This data is small enough that we could probably just use a list and pay the cost of shifting it whenever we need to roll the window. But a more efficient way of doing this, especially if the data window is larger, would be to use a [circular buffer]. For brevity, the code for the circular buffer is not included in this post, but you may find in the [demo code][ddos-detection circularbuffer].
+
+Each server's state needs to keep track of a few things:
+
+1. `address` - the server's address
+2. `window` - a rolling window of per-second data aggregates for requests per second and unique clients per second
+3. `requests_mean` - the weighted mean of the requests data over the window
+4. `requests_stdev` - the weighted standard deviation of the requests data over the window
+5. `clients_mean` - the weighted mean of the clients data over the window
+6. `clients_stdev` - the weighted standard deviation of the clients data over the window
+7. `current_ts` - the current partial second's timestamp
+8. `current_requests` - the number of requests for the current (partial) second
+9. `current_clients` - a set of unique client addresses (to produce the number of clients per second aggregate)
+10. `is_attack` - the current health status of the server
+
+```python
+class SingleServerLoadStatistics(object):
+    def __init__(self, address):
+        self.address = address
+        self.window = CircularBuffer(60)
+        self.requests_mean = 0
+        self.requests_stdev = 0
+        self.clients_mean = 0
+        self.clients_stdev = 0
+        self.current_ts = 0
+        self.current_requests = 0
+        self.current_clients = set()
+        self.is_attack = False
+```
+
+That seems simple enough. But we also need to implement three pieces of logic:
+
+1. Advancing the window and recomputing the means and standard deviations (i.e. updating our model)
```

Comment:
Rather than (i.e. updating our model) I think this reads better as 

1. Updating our model by advancing the window...